File: solutions/csharp/interest-is-interesting/1/InterestIsInteresting.cs
static class SavingsAccount
{
  private static float NegativeBalanceInterestRate = 3.213f;
  private static float LowBalanceInterestRate = 0.5f;
  private static float MediumBalanceInterestRate = 1.621f;
  private static float HighBalanceInterestRate = 2.475f;

  private static int NoBalance = 0;
  private static int LowBalance = 1000;
  private static int HighBalance = 5000;

  public static float InterestRate(decimal balance)
  {
    if (balance < NoBalance) return NegativeBalanceInterestRate;
    if (balance < LowBalance) return LowBalanceInterestRate;
    if (balance < HighBalance) return MediumBalanceInterestRate;

    return HighBalanceInterestRate;
  }

  public static decimal Interest(decimal balance)
  {
    return balance * (decimal)(InterestRate(balance) / 100);
  }

  public static decimal AnnualBalanceUpdate(decimal balance)
  {
    return balance + Interest(balance);
  }

  public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
  {
    int years = 0;
    while (balance < targetBalance)
    {
      balance = AnnualBalanceUpdate(balance);
      years++;
    }

    return years;
  }
}

File: solutions/cpp/interest-is-interesting/1/interest_is_interesting.cpp
#include <iostream>
#include <ostream>
// interest_rate returns the interest rate for the provided balance.
double interest_rate(const double balance) {
    if (balance < 0) {
        return 3.213;
    }if (balance < 1'000) {
        return 0.5;
    }if (balance < 5'000) {
        return 1.621;
    }
    return 2.475;
}

// yearly_interest calculates the yearly interest for the provided balance.
double yearly_interest(const double balance) {
    return balance * (interest_rate(balance)/100.0);
}

// annual_balance_update calculates the annual balance update, taking into
// account the interest rate.
double annual_balance_update(const double balance) {
    return balance + yearly_interest(balance);
}

// years_until_desired_balance calculates the minimum number of years required
// to reach the desired balance.
int years_until_desired_balance(const double balance, const double target_balance) {
    if (std::signbit(balance) != std::signbit(target_balance)) {
        throw std::invalid_argument("Balance will never reach target balance");
    }
    double accumulated_balance = balance;
    int years = 0;
    while (accumulated_balance < target_balance) {
        accumulated_balance = annual_balance_update(accumulated_balance);
        ++years;
    }
    return years;
}